source/common/buffer/buffer_impl.h

```diff
@@ -10,6 +10,43 @@
 namespace Envoy {
 namespace Buffer {
 
+/**
+ * An implementation of BufferFragment where a releasor callback is called when the data is
+ * no longer needed.
+ */
+class BufferFragmentImpl : public BufferFragment {
+public:
+  /**
+   * Creates a new wrapper around the externally owned <data> of size <size>.
+   * The caller must ensure <data> is valid until releasor() is called, or for the lifetime of the
+   * fragment. releasor() is called with <data>, <size> and <this> to allow caller to delete
+   * the fragment object.
+   * @param data external data to reference
+   * @param size size of data
+   * @param releasor a callback function to be called when data is no longer needed.
+   */
+  BufferFragmentImpl(const void* data, size_t size,
+                     std::function<void(const void*, size_t, const BufferFragmentImpl*)> releasor)
+      : data_(data), size_(size), releasor_(releasor) {}
+
+  const void* data() const override { return data_; }
+  size_t size() const override { return size_; }
+
+  void done() override {
+    if (releasor_) {
+      releasor_(data_, size_, this);
+    }
+  }
+
+  BufferFragmentImpl(const BufferFragmentImpl&) = delete;
```

Comment:
nit: You can use the existing `NonCopyable` mixin class in the code for this if you want (search code).

Reply:
Done. This is nice.